File: desktop/native/SurveyController.App/Views/AISettingsDialog.cpp
#include "pch.h"
#include "AISettingsDialog.h"
#include "Services/DialogStyling.h"
#include "Services/JsonHelpers.h"
#include "Services/RpcServices.h"
#include "Services/ShellSettings.h"

#include <memory>

namespace winrt::SurveyController::App::Views
{
    using namespace Microsoft::UI::Xaml;
    using namespace Microsoft::UI::Xaml::Automation;
    using namespace Microsoft::UI::Xaml::Controls;
    using namespace Windows::Data::Json;

    namespace
    {
        hstring SelectedTag(RadioButtons const& buttons, hstring const& fallback)
        {
            auto item = buttons.SelectedItem().try_as<RadioButton>();
            return item ? unbox_value_or<hstring>(item.Tag(), fallback) : fallback;
        }

        void SelectTag(RadioButtons const& buttons, hstring const& value)
        {
            for (uint32_t index = 0; index < buttons.Items().Size(); ++index)
            {
                auto item = buttons.Items().GetAt(index).try_as<RadioButton>();
                if (item && unbox_value_or<hstring>(item.Tag(), L"") == value) { buttons.SelectedIndex(static_cast<int32_t>(index)); return; }
            }
            buttons.SelectedIndex(0);
        }
    }

    Windows::Foundation::IAsyncOperation<bool> ShowAISettingsDialogAsync(XamlRoot const& root)
    {
        JsonObject settings;
        hstring parseError;
        if (!Services::TryParseJsonObject(Services::ShellSettings::Current().Json(), settings, parseError)) co_return false;
        auto profile = settings.GetNamedObject(L"aiProfile", JsonObject{});

        auto panel = StackPanel{};
        panel.Spacing(12);
        auto mode = RadioButtons{};
        mode.Header(box_value(L"AI 模式"));
        AutomationProperties::SetName(mode, L"AI 模式");
        auto free = RadioButton{}; free.Content(box_value(L"限时免费")); free.Tag(box_value(L"free"));
        auto providerMode = RadioButton{}; providerMode.Content(box_value(L"自定义服务")); providerMode.Tag(box_value(L"provider"));
        mode.Items().Append(free); mode.Items().Append(providerMode);
        auto provider = RadioButtons{};
        provider.Header(box_value(L"服务商"));
        AutomationProperties::SetName(provider, L"AI 服务商");
        auto deepseek = RadioButton{}; deepseek.Content(box_value(L"DeepSeek")); deepseek.Tag(box_value(L"deepseek"));
        auto custom = RadioButton{}; custom.Content(box_value(L"OpenAI 兼容")); custom.Tag(box_value(L"custom"));
        provider.Items().Append(deepseek); provider.Items().Append(custom);
        auto baseUrl = TextBox{}; baseUrl.Header(box_value(L"接口地址")); baseUrl.Text(profile.GetNamedString(L"baseURL", L"")); AutomationProperties::SetName(baseUrl, L"AI 接口地址");
        auto model = TextBox{}; model.Header(box_value(L"模型")); model.Text(profile.GetNamedString(L"model", L"")); AutomationProperties::SetName(model, L"AI 模型");
        auto hasSavedKey = profile.GetNamedBoolean(L"hasAPIKey", false);
        auto apiKey = PasswordBox{}; apiKey.Header(box_value(L"API 密钥")); apiKey.PlaceholderText(hasSavedKey ? L"已安全保存，留空不修改" : L"输入 API Key"); apiKey.PasswordRevealMode(PasswordRevealMode::Peek); AutomationProperties::SetName(apiKey, L"AI API 密钥");
        auto test = Button{};
        auto testContent = StackPanel{}; testContent.Orientation(Orientation::Horizontal); testContent.Spacing(8);
        auto testIcon = SymbolIcon{}; testIcon.Symbol(Symbol::Refresh);
        auto testLabel = TextBlock{}; testLabel.Text(L"测试连接");
        testContent.Children().Append(testIcon); testContent.Children().Append(testLabel); test.Content(testContent);
        AutomationProperties::SetName(test, L"测试 AI 连接");
        auto status = InfoBar{}; status.IsOpen(false); status.IsClosable(false);
        AutomationProperties::SetLiveSetting(status,
            Microsoft::UI::Xaml::Automation::Peers::AutomationLiveSetting::Assertive);
        panel.Children().Append(mode); panel.Children().Append(provider); panel.Children().Append(baseUrl); panel.Children().Append(model); panel.Children().Append(apiKey); panel.Children().Append(test); panel.Children().Append(status);
        SelectTag(mode, profile.GetNamedString(L"mode", L"free"));
        SelectTag(provider, profile.GetNamedString(L"provider", L"deepseek"));

        auto updateVisibility = [=]()
        {
            auto customMode = SelectedTag(mode, L"free") == L"provider";
            provider.Visibility(customMode ? Visibility::Visible : Visibility::Collapsed);
            baseUrl.Visibility(customMode && SelectedTag(provider, L"deepseek") == L"custom" ? Visibility::Visible : Visibility::Collapsed);
            model.Visibility(customMode ? Visibility::Visible : Visibility::Collapsed);
            apiKey.Visibility(customMode ? Visibility::Visible : Visibility::Collapsed);
            test.Visibility(customMode ? Visibility::Visible : Visibility::Collapsed);
        };
        auto modeRevoker = mode.SelectionChanged(auto_revoke, [updateVisibility](auto const&, auto const&) { updateVisibility(); });
        auto providerRevoker = provider.SelectionChanged(auto_revoke, [updateVisibility](auto const&, auto const&) { updateVisibility(); });
        updateVisibility();

        auto testRevoker = test.Click(auto_revoke, [=](auto const& sender, auto const&) -> fire_and_forget
        {
            auto button = sender.template try_as<Button>();
            if (button) button.IsEnabled(false);
            status.IsOpen(true);
            if (!apiKey.Password().empty())
            {
                status.Severity(InfoBarSeverity::Warning);
                status.Title(L"请先保存 API 密钥");
                status.Message(L"现有连接测试接口只读取安全存储中的密钥。保存后可重新打开并测试。");
                if (button) button.IsEnabled(true);
                co_return;
            }
            status.Severity(InfoBarSeverity::Informational);
            status.Title(L"正在测试连接");
            JsonObject candidate;
            candidate.SetNamedValue(L"mode", JsonValue::CreateStringValue(SelectedTag(mode, L"free")));
            candidate.SetNamedValue(L"provider", JsonValue::CreateStringValue(SelectedTag(provider, L"deepseek")));
            candidate.SetNamedValue(L"baseURL", JsonValue::CreateStringValue(baseUrl.Text()));
            candidate.SetNamedValue(L"model", JsonValue::CreateStringValue(model.Text()));
            try
            {
                auto response = co_await Services::TaskService{}.TestAiAsync(candidate);
                JsonObject state;
                hstring error;
                if (!Services::TryParseJsonObject(response, state, error)) throw hresult_error(E_FAIL, error);
                auto success = state.GetNamedBoolean(L"success", false);
                status.Severity(success ? InfoBarSeverity::Success : InfoBarSeverity::Error);
                status.Title(success ? L"AI 连接正常" : L"AI 连接失败");
                status.Message(state.GetNamedString(L"message", L""));
            }
            catch (hresult_error const& error)
            {
                status.Severity(InfoBarSeverity::Error);
                status.Title(L"AI 连接失败");
                status.Message(error.message());
            }
            catch (std::exception const& error)
            {
                status.Severity(InfoBarSeverity::Error);
                status.Title(L"AI 连接失败");
                status.Message(to_hstring(error.what()));
            }
            catch (...)
            {
                status.Severity(InfoBarSeverity::Error);
                status.Title(L"AI 连接失败");
                status.Message(L"连接测试出现未知错误。");
            }
            if (button) button.IsEnabled(true);
        });

        ContentDialog dialog;
        auto revoker = Services::PrepareContentDialog(dialog, root);
        AutomationProperties::SetName(dialog, L"AI 服务设置");
        dialog.Title(box_value(L"AI 服务设置"));
        dialog.Content(panel);
        dialog.PrimaryButtonText(L"保存");
        dialog.CloseButtonText(L"取消");
        dialog.DefaultButton(ContentDialogButton::Primary);
        auto saved = std::make_shared<bool>(false);
        auto primaryRevoker = dialog.PrimaryButtonClick(auto_revoke,
            [=](ContentDialog const& sender, ContentDialogButtonClickEventArgs const& eventArgs) -> fire_and_forget
        {
            auto dialogCopy = sender;
            auto args = eventArgs;
            auto deferral = args.GetDeferral();
            dialogCopy.IsPrimaryButtonEnabled(false);
            try
            {
                auto updatedProfile = JsonObject::Parse(profile.Stringify());
                updatedProfile.SetNamedValue(L"mode", JsonValue::CreateStringValue(SelectedTag(mode, L"free")));
                updatedProfile.SetNamedValue(L"provider", JsonValue::CreateStringValue(SelectedTag(provider, L"deepseek")));
                updatedProfile.SetNamedValue(L"baseURL", JsonValue::CreateStringValue(baseUrl.Text()));
                updatedProfile.SetNamedValue(L"model", JsonValue::CreateStringValue(model.Text()));
                if (!updatedProfile.HasKey(L"apiProtocol")) updatedProfile.SetNamedValue(L"apiProtocol", JsonValue::CreateStringValue(L"auto"));
                auto updatedSettings = JsonObject::Parse(settings.Stringify());
                updatedSettings.SetNamedValue(L"aiProfile", updatedProfile);
                JsonObject credential;
                credential.SetNamedValue(L"operation", JsonValue::CreateStringValue(apiKey.Password().empty() ? L"keep" : L"replace"));
                if (!apiKey.Password().empty()) credential.SetNamedValue(L"apiKey", JsonValue::CreateStringValue(apiKey.Password()));
                JsonObject request;
                request.SetNamedValue(L"settings", updatedSettings);
                request.SetNamedValue(L"aiCredential", credential);
                auto response = co_await Services::SettingsService{}.SaveAsync(request.Stringify());
                Services::ShellSettings::Current().Update(response);
                *saved = true;
            }
            catch (hresult_error const& error)
            {
                args.Cancel(true);
                status.Severity(InfoBarSeverity::Error);
                status.Title(L"AI 设置保存失败");
                status.Message(error.message());
                status.IsOpen(true);
            }
            catch (std::exception const& error)
            {
                args.Cancel(true);
                status.Severity(InfoBarSeverity::Error);
                status.Title(L"AI 设置保存失败");
                status.Message(to_hstring(error.what()));
                status.IsOpen(true);
            }
            catch (...)
            {
                args.Cancel(true);
                status.Severity(InfoBarSeverity::Error);
                status.Title(L"AI 设置保存失败");
                status.Message(L"保存设置时出现未知错误。");
                status.IsOpen(true);
            }
            dialogCopy.IsPrimaryButtonEnabled(true);
            deferral.Complete();
        });
        auto result = co_await dialog.ShowAsync();
        co_return result == ContentDialogResult::Primary && *saved;
    }
}

File: desktop/native-cs/src/SurveyController.App/Views/AISettingsDialog.cs
using System.Text.Json.Nodes;
using Microsoft.UI.Xaml;
using Microsoft.UI.Xaml.Automation;
using Microsoft.UI.Xaml.Controls;
using SurveyController.App.Services;
using SurveyController.Core.Settings;

namespace SurveyController.App.Views;

/// <summary>
/// AI 服务设置弹层：全部控件以代码构建，行为对照 C++ Views/AISettingsDialog.cpp。
/// 返回值表示是否保存成功并关闭。
/// </summary>
internal static class AISettingsDialog
{
    public static async Task<bool> ShowAsync(XamlRoot root)
    {
        if (JsonNode.Parse(ShellSettings.Current.Json) is not JsonObject settings)
        {
            return false;
        }
        var profile = settings["aiProfile"] as JsonObject ?? new JsonObject();

        var panel = new StackPanel { Spacing = 12 };

        var mode = new RadioButtons { Header = "AI 模式" };
        AutomationProperties.SetName(mode, "AI 模式");
        mode.Items.Add(new RadioButton { Content = "限时免费", Tag = "free" });
        mode.Items.Add(new RadioButton { Content = "自定义服务", Tag = "provider" });

        var provider = new RadioButtons { Header = "服务商" };
        AutomationProperties.SetName(provider, "AI 服务商");
        provider.Items.Add(new RadioButton { Content = "DeepSeek", Tag = "deepseek" });
        provider.Items.Add(new RadioButton { Content = "OpenAI 兼容", Tag = "custom" });

        var baseUrl = new TextBox { Header = "接口地址", Text = Str(profile, "baseURL", string.Empty) };
        AutomationProperties.SetName(baseUrl, "AI 接口地址");
        var model = new TextBox { Header = "模型", Text = Str(profile, "model", string.Empty) };
        AutomationProperties.SetName(model, "AI 模型");

        var hasSavedKey = Bool(profile, "hasAPIKey");
        var apiKey = new PasswordBox
        {
            Header = "API 密钥",
            PlaceholderText = hasSavedKey ? "已安全保存，留空不修改" : "输入 API Key",
            PasswordRevealMode = PasswordRevealMode.Peek,
        };
        AutomationProperties.SetName(apiKey, "AI API 密钥");

        var testLabel = new TextBlock { Text = "测试连接" };
        var testContent = new StackPanel { Orientation = Orientation.Horizontal, Spacing = 8 };
        testContent.Children.Add(new SymbolIcon(Symbol.Refresh));
        testContent.Children.Add(testLabel);
        var test = new Button { Content = testContent };
        AutomationProperties.SetName(test, "测试 AI 连接");

        var status = new InfoBar { IsOpen = false, IsClosable = false };
        AutomationProperties.SetLiveSetting(status, Microsoft.UI.Xaml.Automation.Peers.AutomationLiveSetting.Assertive);

        panel.Children.Add(mode);
        panel.Children.Add(provider);
        panel.Children.Add(baseUrl);
        panel.Children.Add(model);
        panel.Children.Add(apiKey);
        panel.Children.Add(test);
        panel.Children.Add(status);

        SelectTag(mode, Str(profile, "mode", "free"));
        SelectTag(provider, Str(profile, "provider", "deepseek"));

        void UpdateVisibility()
        {
            var customMode = SelectedTag(mode, "free") == "provider";
            provider.Visibility = customMode ? Visibility.Visible : Visibility.Collapsed;
            baseUrl.Visibility = customMode && SelectedTag(provider, "deepseek") == "custom"
                ? Visibility.Visible
                : Visibility.Collapsed;
            model.Visibility = customMode ? Visibility.Visible : Visibility.Collapsed;
            apiKey.Visibility = customMode ? Visibility.Visible : Visibility.Collapsed;
            test.Visibility = customMode ? Visibility.Visible : Visibility.Collapsed;
        }

        mode.SelectionChanged += (_, _) => UpdateVisibility();
        provider.SelectionChanged += (_, _) => UpdateVisibility();
        UpdateVisibility();

        test.Click += async (_, _) =>
        {
            test.IsEnabled = false;
            status.IsOpen = true;
            if (apiKey.Password.Length > 0)
            {
                status.Severity = InfoBarSeverity.Warning;
                status.Title = "请先保存 API 密钥";
                status.Message = "现有连接测试接口只读取安全存储中的密钥。保存后可重新打开并测试。";
                test.IsEnabled = true;
                return;
            }

            status.Severity = InfoBarSeverity.Informational;
            status.Title = "正在测试连接";
            var candidate = new JsonObject
            {
                ["mode"] = SelectedTag(mode, "free"),
                ["provider"] = SelectedTag(provider, "deepseek"),
                ["baseURL"] = baseUrl.Text,
                ["model"] = model.Text,
            };
            try
            {
                var response = await TaskService.TestAiConnectionAsync(candidate.ToJsonString());
                if (JsonNode.Parse(response) is not JsonObject state)
                {
                    throw new InvalidOperationException("后端响应格式无效");
                }
                var success = Bool(state, "success");
                status.Severity = success ? InfoBarSeverity.Success : InfoBarSeverity.Error;
                status.Title = success ? "AI 连接正常" : "AI 连接失败";
                status.Message = Str(state, "message", string.Empty);
            }
            catch (Exception error)
            {
                status.Severity = InfoBarSeverity.Error;
                status.Title = "AI 连接失败";
                status.Message = error.Message;
            }
            test.IsEnabled = true;
        };

        var dialog = new ContentDialog
        {
            Title = "AI 服务设置",
            Content = panel,
            PrimaryButtonText = "保存",
            CloseButtonText = "取消",
            DefaultButton = ContentDialogButton.Primary,
        };
        AutomationProperties.SetName(dialog, "AI 服务设置");
        DialogStyling.PrepareContentDialog(dialog, root);

        var saved = false;
        dialog.PrimaryButtonClick += async (sender, args) =>
        {
            var deferral = args.GetDeferral();
            sender.IsPrimaryButtonEnabled = false;
            try
            {
                // 克隆既有 profile/settings，避免污染当前快照（同 C++ Parse(Stringify())）。
                var updatedProfile = JsonNode.Parse(profile.ToJsonString())!.AsObject();
                updatedProfile["mode"] = SelectedTag(mode, "free");
                updatedProfile["provider"] = SelectedTag(provider, "deepseek");
                updatedProfile["baseURL"] = baseUrl.Text;
                updatedProfile["model"] = model.Text;
                if (!updatedProfile.ContainsKey("apiProtocol"))
                {
                    updatedProfile["apiProtocol"] = "auto";
                }
                var updatedSettings = JsonNode.Parse(settings.ToJsonString())!.AsObject();
                updatedSettings["aiProfile"] = updatedProfile;

                var credential = new JsonObject
                {
                    ["operation"] = apiKey.Password.Length > 0 ? "replace" : "keep",
                };
                if (apiKey.Password.Length > 0)
                {
                    credential["apiKey"] = apiKey.Password;
                }
                var request = new JsonObject
                {
                    ["settings"] = updatedSettings,
                    ["aiCredential"] = credential,
                };

                var response = await SettingsService.SaveAsync(request.ToJsonString());
                ShellSettings.Current.Update(response);
                saved = true;
            }
            catch (Exception error)
            {
                args.Cancel = true;
                status.Severity = InfoBarSeverity.Error;
                status.Title = "AI 设置保存失败";
                status.Message = error.Message;
                status.IsOpen = true;
            }
            sender.IsPrimaryButtonEnabled = true;
            deferral.Complete();
        };

        var result = await dialog.ShowAsync();
        return result == ContentDialogResult.Primary && saved;
    }

    private static string SelectedTag(RadioButtons buttons, string fallback) =>
        buttons.SelectedItem is RadioButton item && item.Tag is string tag ? tag : fallback;

    private static void SelectTag(RadioButtons buttons, string value)
    {
        for (var index = 0; index < buttons.Items.Count; index++)
        {
            if (buttons.Items[index] is RadioButton item && item.Tag as string == value)
            {
                buttons.SelectedIndex = index;
                return;
            }
        }
        buttons.SelectedIndex = 0;
    }

    private static string Str(JsonObject parent, string name, string fallback) =>
        parent[name] is JsonValue value && value.TryGetValue<string>(out var text) ? text : fallback;

    private static bool Bool(JsonObject parent, string name, bool fallback = false) =>
        parent[name] is JsonValue value && value.TryGetValue<bool>(out var flag) ? flag : fallback;
}
